tests/operators/_test_offres_exp_approx.py

```diff
@@ -0,0 +1,236 @@
+
+if AUTOGRAD_AVAILABLE:
+    params.extend([(2, "torch", "cpu"), (3, "torch", "cpu")])
+
+if AUTOGRAD_AVAILABLE and CUPY_AVAILABLE:
+    params.extend([(2, "torch", "cuda"), (3, "torch", "cuda")])
+
+
+def calculate_true_offresonance_term(fieldmap, t):
+    """Calculate non-approximate off-resonance modulation term."""
+    xp = get_array_module(fieldmap)
+    t = xp.asarray(t)
+    if xp.__name__ == "torch":
+        t = t.to(device=fieldmap.device)
+    arg = t * fieldmap[..., None]
+    arg = arg[None, ...].swapaxes(0, -1)[..., 0]
+    return xp.exp(-arg)
+
+
+def calculate_approx_offresonance_term(B, C):
+    """Calculate approximate off-resonance modulation term."""
+    field_term = 0.0
+    for n in range(B.shape[0]):
+        tmp = B[n] * C[n][..., None]
+        tmp = tmp[None, ...].swapaxes(0, -1)[..., 0]
+        field_term += tmp
+
+    return field_term
+
+
+def make_b0map(ndim, npix=64):
+    """Make B0 map (units: Hz)."""
+    # generate support
+    shape = ndim * [npix]
+    if ndim == 2:
+        mask, fieldmap = _make_disk(shape)
+    elif ndim == 3:
+        mask, fieldmap = _make_sphere(shape)
+
+    # mask map
+    fieldmap *= mask
+
+    # rescale
+    fieldmap = 600 * fieldmap / fieldmap.max() - 300  # Hz
+
+    # mask map
+    fieldmap *= mask
+
+    return fieldmap.astype(np.float32), mask
+
+
+def make_t2smap(ndim, npix=64):
+    """Make T2* map (units: ms)."""
+    shape = ndim * [npix]
+    if ndim == 2:
+        mask, _ = _make_disk(shape)
+    elif ndim == 3:
+        mask, _ = _make_sphere(shape)
+
+    # rescale
+    fieldmap = 15.0 * mask  # ms
+
+    return fieldmap.astype(np.float32), mask
+
+
+def _make_disk(shape, frac_radius=0.3):
+    """Make circular binary mask."""
+    # calculate grid
+    ny, nx = shape
+    yy, xx = np.mgrid[:ny, :nx]
+    yy, xx = yy - ny // 2, xx - nx // 2
+    yy, xx = yy / ny, xx / nx
+
+    # radius
+    rr = (xx**2 + yy**2) ** 0.5
+    return rr < frac_radius, rr
+
+
+def _make_sphere(shape, frac_radius=0.3):
+    """Make spherical binary mask."""
+    # calculate grid
+    nz, ny, nx = shape
+    zz, yy, xx = np.mgrid[:nz, :ny, :nx]
+    zz, yy, xx = zz - nz // 2, yy - ny // 2, xx - nx // 2
+    zz, yy, xx = zz / nz, yy / ny, xx / nx
+
+    # radius
+    rr = (xx**2 + yy**2 + zz**2) ** 0.5
+    return rr < frac_radius, rr
+
+
+def get_backend(backend):
+    if backend == "cpu":
+        return np
+    elif backend == "gpu":
+        return cp
+    elif backend == "torch":
+        return torch
+
+
+def assert_allclose(actual, expected, atol, rtol):
+    """Backend agnostic assertion."""
+    xp = get_array_module(actual)
+    if xp.__name__ == "numpy":
+        npt.assert_allclose(actual, expected, atol, rtol)
+    if xp.__name__ == "cupy":
+        npt.assert_allclose(actual.get(), expected.get(), atol, rtol)
+    if xp.__name__ == "torch":
+        npt.assert_allclose(
+            actual.numpy(force=True), expected.numpy(force=True), atol, rtol
```

Comment:
you can instead use the `from_interface` helper, directly in the test. otherwise, if the test is failing you don't get a nice traceback